src/config/mongoose_config_parser_toml.erl

```diff
@@ -0,0 +1,354 @@
+    State = mongoose_config_parser:opts_to_state(Opts ++ HOpts, Hosts),
+    State1 = mongoose_config_parser:dedup_state_opts(State),
+    mongoose_config_parser:add_dep_modules(State1).
+
+parse(Content) ->
+    lists:append([process_section(Section, SecContent) ||
+                     {Section, SecContent} <- maps:to_list(Content)]).
+
+process_section(<<"acl">>, Content) ->
+    [process_acl(Name, global, C) || {Name, C} <- maps:to_list(Content)];
+process_section(<<"access">>, Content) ->
+    [process_access_rule(Name, global, C) || {Name, C} <- maps:to_list(Content)];
+process_section(<<"shaper">>, Content) ->
+    [process_shaper(Name, global, C) || {Name, C} <- maps:to_list(Content)];
+process_section(<<"outgoing_pools">>, Content) ->
+    Pools = lists:append([process_pool_type(Type, C) || {Type, C} <- maps:to_list(Content)]),
+    [#local_config{key = outgoing_pools, value = Pools}];
+process_section(<<"host_config">>, Content) ->
+    lists:append([process_host_item(C) || C <- Content]);
+process_section(<<"listen">>, Content) ->
+    Listeners = lists:append([process_listener_type(Type, C)
+                              || {Type, C} <- maps:to_list(Content)]),
+    [#local_config{key = listen, value = Listeners}];
+process_section(<<"general">>, Content) ->
+    [process_general(Name, C) || {Name, C} <- maps:to_list(Content)];
+process_section(<<"s2s">>, Content) ->
+    lists:append([process_s2s_option(Name, C) || {Name, C} <- maps:to_list(Content)]);
+process_section(<<"auth">>, Content = #{<<"methods">> := Methods}) ->
+    AuthOpts = auth_opts(maps:without([<<"methods">>], Content)),
+    [?HOST_F(#local_config{key = {auth_method, Host},
+                           value = [b2a(Method) || Method <- Methods]}),
+     ?HOST_F(#local_config{key = {auth_opts, Host},
+                           value = [Opt || {inner, Opt} <- AuthOpts]}) |
+     [?HOST_F(#local_config{key = {K, Host}, value = V}) || {outer, {K, V}} <- AuthOpts]];
+process_section(<<"modules">>, Content) ->
+    Mods = [process_module(Mod, Opts) || {Mod, Opts} <- maps:to_list(Content)],
+    [?HOST_F(#local_config{key = {modules, Host}, value = Mods})];
+process_section(<<"services">>, Content) ->
+    Services = [process_service(S, Opts) || {S, Opts} <- maps:to_list(Content)],
+    [#local_config{key = services, value = Services}].
+
+process_module(Mod, Opts) ->
+    {b2a(Mod), [module_opt(Mod, K, V) || {K, V} <- maps:to_list(Opts)]}.
+
+module_opt(<<"mod_disco">>, <<"users_can_see_hidden_services">>, V) ->
+    {users_can_see_hidden_services, V};
+module_opt(<<"mod_register">>, <<"welcome_message">>, V) ->
+    {welcome_message, b2l(V)};
+module_opt(<<"mod_register">>, <<"ip_access">>, V) ->
+    Rules = [{Addr, b2a(P)} || #{<<"address">> := Addr, <<"policy">> := P} <- V],
+    {ip_access, Rules};
+module_opt(<<"mod_register">>, <<"access">>, V) ->
+    {access, b2a(V)};
+module_opt(<<"mod_vcard">>, <<"host">>, V) ->
+    {host, b2l(V)};
```

Comment:
I think this kind of parsing could be delegated to the respective modules so they control their specific config options. Probably in a separate PR.

Reply:
Yes I think it needs to be well thought and probably aligned with the work around the config API you are starting now.